src/tree.rs

```diff
@@ -101,3 +101,138 @@ fn hash_layer(l_star: usize, pair: Pair) -> pallas::Base {
+            let domain = HashDomain::new(MERKLE_CRH_PERSONALIZATION);
+
+            let mut v = vec![pallas::Base::zero()];
+            for l_star in 0..MERKLE_DEPTH_ORCHARD {
+                let next = domain
+                .hash(
+                    std::iter::empty()
+                        .chain(i2lebsp_k(l_star).iter().copied().take(K))
+                        .chain(
+                            v[l_star]
+                            .to_le_bits()
+                            .iter()
+                            .by_val()
+                            .take(pallas::Base::NUM_BITS as usize)
+                        )
+                        .chain(
+                            v[l_star]
+                            .to_le_bits()
+                            .iter()
+                            .by_val()
+                            .take(pallas::Base::NUM_BITS as usize)
+                        ),
+                )
+                .unwrap();
+                v.push(next);
+            }
+            v
+        };
+    }
+
+    prop_compose! {
+        /// Generates an arbitrary Merkle tree of with `n_notes` nonempty leaves.
+        pub fn arb_tree(n_notes: i32)
+        (
+            // generate note values that we're certain won't exceed MAX_NOTE_VALUE in total
+            notes in vec(
+                arb_positive_note_value(MAX_NOTE_VALUE / n_notes as u64).prop_flat_map(arb_note),
+                n_notes as usize
+            ),
+        )
+        -> (Vec<(Note, MerklePath)>, Anchor) {
+            // Inefficient algorithm to build a perfect subtree containing all notes.
+            let perfect_subtree_depth = (n_notes as f64).log2().ceil() as usize;
+            let commitments: Vec<ExtractedNoteCommitment> = notes.iter().map(|note| {
+                let cmx: ExtractedNoteCommitment = note.commitment().into();
+                cmx
+            }).collect();
+
+            let padded_leaves = {
+                let mut padded_leaves = commitments.clone();
+
+                let pad = (0..((1 << perfect_subtree_depth) - n_notes as usize)).map(
+                    |_| ExtractedNoteCommitment::uncommitted()
+                ).collect::<Vec<_>>();
+
+                padded_leaves.extend_from_slice(&pad);
+                padded_leaves
+            };
+
+            let tree = {
+                let mut tree: Vec<Vec<pallas::Base>> = vec![padded_leaves.into_iter().map(|leaf| *leaf).collect()];
+                for height in 1..perfect_subtree_depth {
+                    let inner_nodes = (0..(perfect_subtree_depth >> height)).map(|pos| {
+                        hash_layer(height, Pair {
+                            left: tree[height - 1][pos * 2],
+                            right: tree[height - 1][pos * 2 + 1],
+                        })
+                    }).collect();
+                    tree.push(inner_nodes);
+                };
+                tree
+            };
```

Comment:
Then in here:
- When both left and right are `None`, push `None` (skipping the `hash_layer`, which was already pre-computed in `EMPTY_ROOTS`).
- When either one left or right is `None`, fetch the corresponding value from `EMPTY_ROOTS`.

In the auth path logic below, values in `tree` can just be unwrapped, since all required siblings in the tree will have been filled in.